src/memory/combat_manager.rs

```diff
@@ -51,4 +143,183 @@ impl CombatManagerData {
 
         Ok(())
     }
+
+    pub fn update_combat_controller_type(
+        &mut self,
+        memory_context: &MemoryContext,
+    ) -> Result<(), MemoryError> {
+        // [self.current_encounter_base, 0x128, 0x0, 0x10, 0x0],
+        if let Ok(controller) =
+            memory_context.follow_fields::<u8>(&["currentEncounter", "controller"])
+        {
+            // This code reaches into the base types of the controller thats active to find the
+            // name
+            if let Ok(controller_type_c_str) = memory_context
+                .read_pointer_path::<ArrayCString<200>>(&[controller.into(), 0x0, 0x10, 0x0])
+            {
+                if let Ok(controller_type) = controller_type_c_str.validate_utf8() {
+                    self.combat_controller_type = match controller_type {
+                        "EncounterController" => CombatControllerType::Basic,
+                        "FirstEncounter" => CombatControllerType::FirstEncounter,
+                        "SecondEncounter" => CombatControllerType::SecondEncounter,
+                        "DwellerOfStrife" => CombatControllerType::DwellerOfStrife,
+                        "DwellerOfDread" => CombatControllerType::DwellerOfDread,
+                        "KOTutorial" => CombatControllerType::KOTutorial,
+                        "LiveManaTutorial" => CombatControllerType::LiveManaTutorial,
+                        "ManaRegenTutorial" => CombatControllerType::ManaRegenTutorial,
+                        "RoundsTutorial" => CombatControllerType::RoundsTutorial,
+                        "SpellLockTutorial" => CombatControllerType::SpellLockTutorial,
+                        "TimedBlocksTutorial" => CombatControllerType::TimedBlocksTutorial,
+                        "TimedHitsTutorial" => CombatControllerType::TimedHitsTutorial,
+                        _ => CombatControllerType::Basic,
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn update_live_mana(&mut self, memory_context: &MemoryContext) -> Result<(), MemoryError> {
+        if let Ok(small_live_mana_ptr) = memory_context.follow_fields::<u64>(&[
+            "currentEncounter",
+            "liveManaHandler",
+            "smallLiveManaParticles",
+        ]) {
+            if let Ok(small_mana) =
+                memory_context.read_pointer_path::<u32>(&[small_live_mana_ptr, 0x18])
+            {
+                self.live_mana.small = small_mana
+            }
```

Comment:
No risk for `small_live_mana_ptr` to be `0x0` here? Same with big below.

Reply:
Shouldnt be! Its there as 0 in all part of the early game - worst case the battle ends and the controller isn't usable.